### PR TITLE
[DS][1/n] Create SchedulingConditionEvaluationContext

### PR DESCRIPTION
**Pull request by OwenKephart**
## Summary & Motivation

Creates a new SchedulingConditionEvaluationContext. This is essentially a "fresh start" for rebuilding the context object, to make it easier to refactor / delete the current state of the AssetConditionEvaluationContext in the future. It has a greatly reduced set of properties and methods, which can be incrementally built up as needed.

It does keep around a reference to the legacy AssetConditionEvaluationContext, for the sole purpose of letting us keep the existing AutoMaterializeRule implementations untouched. In the eventual future where these are deleted, we can very easily sever the entire legacy context from our new context object.

Note that the weird Union[Scheduling...Context, Asset...Context] thing will go away upstack, so don't worry about that bit.

### CODE REVIEW

**Comment by OwenKephart** [comment]
* **#21504** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21504?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21503** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21503?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21502** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21502?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21501** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21501?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21500** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21500?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21499** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21499?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @OwenKephart and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/dagster/dagster/_core/definitions/asset_condition/scheduling_condition_evaluation_context.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    AssetConditionEvaluationState,
+)
+from dagster._core.definitions.asset_key import AssetKey
+from dagster._core.definitions.asset_subset import ValidAssetSubset
+from dagster._core.definitions.events import AssetKeyPartitionKey
+
+from .asset_condition_evaluation_context import AssetConditionEvaluationContext
+
+
+@dataclass(frozen=True)
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/_core/definitions/asset_condition/scheduling_condition_evaluation_context.py:25]: Would love to use DagsterModel here, but was running into Pydantic stuff because we use a `PackableValue` type inside AssetConditionEvaluationState, which causes errors due to its weird type signature.

I'll try to fix that up later in the stack, but for now just using dataclass

File: python_modules/dagster/dagster/_core/definitions/asset_condition/asset_condition.py [hunk 2]
```diff
@@ -31,6 +31,7 @@
 
 if TYPE_CHECKING:
     from .asset_condition_evaluation_context import AssetConditionEvaluationContext
+    from .scheduling_condition_evaluation_context import SchedulingConditionEvaluationContext
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/asset_condition/asset_condition.py:34]: We can rename in a sep diff but i think we can shorten this to `SchedulingContext`

File: python_modules/dagster/dagster/_core/definitions/asset_daemon_context.py [hunk 3]
```diff
@@ -101,6 +106,19 @@ def __init__(
...
+        )
+        self._asset_graph_view = AssetGraphView(
+            temporal_context=TemporalContext(
+                effective_dt=self.instance_queryer.evaluation_time,
+                # TODO: we should eventually pass in an explicit last_event_id
+                last_event_id=None,
+            ),
+            stale_resolver=stale_resolver,
+        )
+        self._data_time_resolver = CachingDataTimeResolver(self.instance_queryer)
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/asset_daemon_context.py:121]: What is the future of `CachingDataTimeResolver`? Is it going away with freshness policies? Or does this need to be roped into `AssetGraphView`?

**Review by schrockn — CHANGES_REQUESTED** [review]

File: python_modules/dagster/dagster/_core/definitions/asset_condition/asset_condition.py (hunk 2 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/asset_condition/asset_condition.py:34]: Added a diff upstack to do this rename

File: python_modules/dagster/dagster/_core/definitions/asset_condition/scheduling_condition_evaluation_context.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+class SchedulingConditionEvaluationContext:
+    # the AssetKey of the currently-evaluated asset
+    asset_key: AssetKey
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/asset_condition/scheduling_condition_evaluation_context.py:29]: would like to do a bit of thinking about the asset check key case here

File: python_modules/dagster/dagster/_core/definitions/asset_condition/scheduling_condition_evaluation_context.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+    # the condition that is being evaluated
+    condition: AssetCondition
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/asset_condition/scheduling_condition_evaluation_context.py:32]: When will we introduce `SchedulingCondition`?

File: python_modules/dagster/dagster/_core/definitions/asset_condition/scheduling_condition_evaluation_context.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    # the subset of AssetPartitions for this Asset which must be evaluated. at the root of the
+    # condition evaluation tree, this is the AllPartitionsSubset, but this may shrink as conditions
+    # are applied and we can be certain that certain partitions will not be true for the overall
+    # expression
+    candidate_subset: ValidAssetSubset
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/asset_condition/scheduling_condition_evaluation_context.py:37]: let's do `AssetSlice`.

Also I think we need to think asset check scheduling here.

File: python_modules/dagster/dagster/_core/definitions/asset_condition/scheduling_condition_evaluation_context.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+    # a view of the AssetGraph that will be used to help compute properties of the asset during
+    # computation
+    asset_graph_view: AssetGraphView
+
+    # the serialized information calculated during the previous evaluation of this condition
+    # note that this refers to the evaluation of this specific node in the condition tree, not the
+    # evaluation of the root of the tree
+    # used to avoid recomputing information that we know has not changed since the previous tick
+    previous_evaluation: Optional[AssetConditionEvaluation]
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/asset_condition/scheduling_condition_evaluation_context.py:47]: I think we should build up a new evaluation object from scratch

Comment (schrockn) [reply python_modules/dagster/dagster/_core/definitions/asset_condition/scheduling_condition_evaluation_context.py:47]: (Even if it contains an AssetConditionEvaluation object internally)

**Review by schrockn — CHANGES_REQUESTED** [review]
Per offline discussion, we should not commit too much in this PR

1) Separately explore what buildling a standalone scheduilng condition on an asset check looks like. That will need to co-exist in the same system
2) Don't expose stuff like `AssetConditionEvaluation`. We should consider that a hidden, storage-only class and instead have different APIs that are designed to make life of the product engineer building a rule as easy as possible. 

An example is that `candidate_subset: Union[AssetSubset, HistoricalAllPartitionsSubsetSentinel]` should be abstracted away from the public interface.

File: python_modules/dagster/dagster/_core/definitions/asset_condition/scheduling_condition_evaluation_context.py (hunk 4 above)
Comment (schrockn) [reply python_modules/dagster/dagster/_core/definitions/asset_condition/scheduling_condition_evaluation_context.py:29]: Just checking back in here. @OwenKephart I merged this entire stack [here](https://github.com/dagster-io/dagster/pull/21582) and looked through it and I really like the direction it is heading. So I am largely into this stack going skip pushing back stuff that I know will nor survive the stack, even if I would object to it normally :-)

File: python_modules/dagster/dagster/_core/definitions/asset_condition/scheduling_condition_evaluation_context.py (hunk 6 above)
Comment (schrockn) [reply python_modules/dagster/dagster/_core/definitions/asset_condition/scheduling_condition_evaluation_context.py:37]: This is changed by the end of this stack so letting it go

**Review by schrockn — APPROVED** [review]

**Comment by OwenKephart** [comment]
### Merge activity

* **May 3, 5:31 PM EDT**: @OwenKephart started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/21499).
